src/factorizations.jl

```diff
@@ -0,0 +1,75 @@
+using BlockArrays: blocks
+using BlockSparseArrays:
+  BlockSparseArrays,
+  BlockSparseMatrix,
+  BlockPermutedDiagonalAlgorithm,
+  BlockPermutedDiagonalTruncationStrategy,
+  diagview,
+  eachblockaxis,
+  mortar_axis
+using LinearAlgebra: Diagonal
+using MatrixAlgebraKit: MatrixAlgebraKit, svd_compact!, svd_full!, svd_trunc!
+
+function BlockSparseArrays.similar_output(
+  ::typeof(svd_compact!),
+  A::GradedMatrix,
+  s_axis::AbstractUnitRange,
+  alg::BlockPermutedDiagonalAlgorithm,
+)
+  u_axis = s_axis
+  flx = flux(A)
+  axs = eachblockaxis(s_axis)
+  # TODO: Use `gradedrange` constructor.
+  v_axis = mortar_axis(
+    map(axs) do ax
+      return sectorrange(dual(sector(ax)) ⊗ flx, ungrade(ax))
+    end,
+  )
+  U = similar(A, axes(A, 1), dual(u_axis))
+  T = real(eltype(A))
+  S = BlockSparseMatrix{T,Diagonal{T,Vector{T}}}(undef, (u_axis, dual(v_axis)))
```

Comment:
I noticed that this is not very general (for example it would be bad on GPU), probably it should be factored out into a `similar_diagonal_blocks` function or something like that.

Reply:
I agree, this is more or less why I had an original function `_similar_S` there to specialize on, since I don't think we really have the infrastructure right now to make that work.
The only thing I can come up with is trying to get type inference to tell us what the output type of `initialize_output` would be on the blocktypes of the input, but that also seems somewhat brittle.

Reply:
Yeah, it would be nice to come up with a good code pattern for this. I think `promote_op` on `initialize_output` of the input blocktypes could be a good general fallback definition, and then we could use `similar(a, BlockType(...))` once we define that.